[Lin Hui Xin Tiffany] iP

Tiffanylin21 opened the pull request:
# Duke
> This is a description of Duke, the CLI task manager

Duke is your pocket task manager, it is:
- text-based
- ~fast~ *super* fast
- easy to learn
All you need to do is: 
1. Download it from [here](https://github.com/Tiffanylin21/ip.git)
2. Add your tasks
3. Let Duke manage your tasks for you!
 And it's **free**!

Features:
- [x] Manage Events, Deadlines and Todos
- [x] Mark tasks as done when complete :tada:
- [ ] Simple and user-friendly GUI for a hassle-free experience (Coming soon)

If you would like to specify a folder and file for storage, please modify the `main` method which looks like this:
```java
public static void main(String[] args) throws IOException {
    String dirPath = "./src/main/data";
    String filePath = "./src/main/data/duke.txt";
    Duke duke = new Duke(dirPath, filePath);
    duke.run();
}
```

Commit:
Add Gradle support

Tiffanylin21 committed:
Implemented skeletal version of Duke (Greet, Echo, Exit)

Tiffanylin21 committed:
Updated skeletal duke

Tiffanylin21 committed:
Implemented Add, List

Tiffanylin21 committed:
Implemented Mark as Done with extension A-Classes

Tiffanylin21 committed:
Implemented ToDo, Event, Deadlines with extension A-Inheritance

Tiffanylin21 committed:
Implemented automated text UI testing

Tiffanylin21 committed:
Implemented Handle Errors

Tiffanylin21 committed:
Implemented Delete

Tiffanylin21 committed:
Implemented Use Enums

Tiffanylin21 committed:
Implement Level 7 Save

Tiffanylin21 committed:
Implement Level 8: Dates and Times

Tiffanylin21 committed:
Merge branch 'branch-Level-8' to master

Tiffanylin21 committed:
Implement use more OOP

Tiffanylin21 committed:
Implement Organize into Packages

Tiffanylin21 committed:
Implement A-JUnit

Tiffanylin21 committed:
Add JavaDoc comments to code.

Tiffanylin21 committed:
Tweak code to comply with coding standard

Tiffanylin21 committed:
Add Find command

Tiffanylin21 committed:
Merge branch 'branch-A-JavaDoc' to master

Tiffanylin21 committed:
Merge branch 'branch-A-CodingStandard' to master

Tiffanylin21 committed:
Clean duke.txt

Tiffanylin21 committed:
Merge branch 'add-gradle-support' into branch-A-Gradle

snss231 commented on src/main/java/duke/helpers/TaskList.java:
```diff
@@ -0,0 +1,222 @@
+ */
+public class TaskList {
+
+    private static ArrayList<Task> tasks;
```
Curious to hear why you chose to set tasks as a class-level member, in my opinion it would make sense for each instance of TaskList to have it's own ArrayList<Task>

snss231 commented on src/main/java/duke/exceptions/InvalidIndexException.java:
```diff
@@ -0,0 +1,7 @@
+package duke.exceptions;
+
+public class InvalidIndexException extends DukeException {
```
Cool that you created subclasses for DukeException :)

snss231 commented on src/main/java/duke/helpers/Parser.java:
```diff
@@ -0,0 +1,97 @@
+     * @return boolean that indicates if the user gave the specified command.
+     * @throws DukeException If deadline and event commands are missing date and/or time.
+     */
+    static boolean isCommand(String s, Command command) throws DukeException {
```
I really like the idea of having static methods for Parser actually, then we can just use Parser like a library of static methods since we don't really need to instantiate multiple Parser objects each with their own state. I might make that change for my own implementation

snss231 reviewed (commented):
Really nice polished implementation 😄 , just shared some thoughts about static members above

snss231 commented on src/main/java/duke/helpers/Storage.java:
```diff
@@ -0,0 +1,119 @@
+    public static ArrayList<Task> saveToFile(ArrayList<Task> taskArr) {
+        try {
+            java.nio.file.Path path = java.nio.file.Paths.get("src/main/data/duke.txt");
+            BufferedWriter writer = java.nio.file.Files.newBufferedWriter(path);
```
In case you weren't aware of this, instead of manually converting the data into text, an alternative way is to use a combination of ObjectInput/OutputStream and FileInput/OutputStream from the Java library to save and load objects that implement Serializable.

Tiffanylin21 committed:
Implement gradle and corrected styling

Tiffanylin21 committed:
Merge branch 'branch-A-Gradle'

laughingkid-sg commented on src/main/java/duke/helpers/Parser.java:
```diff
@@ -0,0 +1,97 @@
+    public static void parse(String input) {
+        String ans = "\t";
+        try {
+            if (isCommand(input, Command.BYE)) {
```
I like the usage of enum in this context. Curious, any reason you use `if/else` block instead of using switch statements here?

laughingkid-sg commented on src/main/java/duke/helpers/Ui.java:
```diff
@@ -0,0 +1,50 @@
+    /**
+     * Closes the Scanner object
+     */
+    public static void closeScanner() {
```
Maybe it will be good to have an exit message implemented here. Not sure if there's any indication of closing the application as I can find it.

laughingkid-sg reviewed (commented):
LGTM. Just some nits to fix.

Tiffanylin21 committed:
Implement Level 10 GUI

Tiffanylin21 committed:
Add assert feature.

Originally, Duke processes empty user inputs.

Asserting for a non-blank user input ensures that Duke will not register
empty inputs which improves user experience when using the
application.

Tiffanylin21 committed:
Add assert feature.

Tiffanylin21 committed:
Improve code quality

Took out commented out code and avoided complicated expressions

Tiffanylin21 committed:
Merge pull request #2 from Tiffanylin21/branch-A-Assertions

Add assert feature.

Tiffanylin21 committed:
Merge branch 'branch-A-Assertions' of ssh://github.com/Tiffanylin21/ip into branch-A-Assertions

Tiffanylin21 committed:
Merge branch 'branch-A-Assertions' into branch-A-CodeQuality

Tiffanylin21 committed:
Merge pull request #3 from Tiffanylin21/branch-A-CodeQuality

Improve code quality

Tiffanylin21 committed:
Improve GUI to make it more polished

Tiffanylin21 committed:
Add representative screenshot of product

Tiffanylin21 committed:
Update README.md

Tiffanylin21 committed:
Set theme jekyll-theme-slate

Tiffanylin21 committed:
Improve coding style

Tiffanylin21 committed:
Update duke GUI

Tiffanylin21 committed:
Update duke GUI

Tiffanylin21 committed:
Update date and time parsing

Tiffanylin21 committed:
Merge branch 'master' of ssh://github.com/Tiffanylin21/ip

Tiffanylin21 committed:
Set theme jekyll-theme-cayman

Tiffanylin21 committed:
Set theme jekyll-theme-midnight

Tiffanylin21 committed:
Set theme jekyll-theme-minimal

Tiffanylin21 committed:
Set theme jekyll-theme-cayman

Tiffanylin21 committed:
Update README.md

Tiffanylin21 committed:
Update README.md

Tiffanylin21 committed:
Add credit

Tiffanylin21 committed:
Merge branch 'master' of ssh://github.com/Tiffanylin21/ip

Tiffanylin21 committed:
Amend JUnit test

Tiffanylin21 committed:
Update README.md

Tiffanylin21 committed:
Update README.md

Tiffanylin21 committed:
Update file paths
